contracts/buy-and-make/RevenueSplitBuyBack.sol

```diff
@@ -0,0 +1,350 @@
+     * minRewardsAmounts = 1 WBTC * slippage 0.97 * MTA decimals 1e18 * MTA/BTC rate 0.00001
+     * = 1 * 0.97 * 1e18 * 0.00001 = 0.97 * 1e13 = 97e11
+     *
+     * @param uniswapPaths The Uniswap V3 bytes encoded paths.
+     */
+    function buyBackRewards(
+        address[] calldata mAssets,
+        uint256[] memory minBassetsAmounts,
+        uint256[] memory minRewardsAmounts,
+        bytes[] calldata uniswapPaths
+    )
+        external
+        onlyKeeperOrGovernor
+        returns (
+            uint256 mAssetToTreasury,
+            uint256 mAssetsSellAmount,
+            uint256 rewardsAmount
+        )
+    {
+        uint256 len = mAssets.length;
+        require(len > 0, "Invalid mAssets");
+        require(minBassetsAmounts.length == len, "Invalid minBassetsAmounts");
+        require(minRewardsAmounts.length == len, "Invalid minRewardsAmounts");
+        require(uniswapPaths.length == len, "Invalid uniswapPaths");
+
+        // for each mAsset
+        for (uint256 i = 0; i < len; i++) {
+            // Get bAsset for mAsset
+            address bAsset = bassets[mAssets[i]];
+            require(bAsset != address(0), "Invalid mAsset");
+            // Validate Uniswap path
+            require(
+                _validUniswapPath(bAsset, address(REWARDS_TOKEN), uniswapPaths[i]),
+                "Invalid uniswap path"
+            );
+
+            // Get mAsset revenue
+            uint256 mAssetBal = IERC20(mAssets[i]).balanceOf(address(this));
+
+            // If a portion of the revenue is being sent to treasury
+            if (treasuryFee > 0) {
+                // STEP 1: Send mAsset to treasury
+                mAssetToTreasury = (mAssetBal * treasuryFee) / CONFIG_SCALE;
+                IERC20(mAssets[i]).safeTransfer(treasury, mAssetToTreasury);
+            }
+
+            // If some portion of the revenue is used to buy back rewards tokens
+            if (treasuryFee < CONFIG_SCALE) {
+                mAssetsSellAmount = mAssetBal - mAssetToTreasury;
+
+                // STEP 2 - Redeem mAssets for bAssets
+                uint256 bAssetAmount = IMasset(mAssets[i]).redeem(
+                    bAsset,
+                    mAssetsSellAmount,
+                    minBassetsAmounts[i],
+                    address(this)
+                );
+
+                // STEP 3 - Swap bAssets for rewards using Uniswap V3
+                IERC20(bAsset).safeApprove(address(UNISWAP_ROUTER), bAssetAmount);
+                IUniswapV3SwapRouter.ExactInputParams memory param = IUniswapV3SwapRouter
+                .ExactInputParams(
+                    uniswapPaths[i],
+                    address(this),
+                    block.timestamp,
+                    bAssetAmount,
+                    minRewardsAmounts[i]
+                );
+                rewardsAmount = UNISWAP_ROUTER.exactInput(param);
+            }
+
+            emit BuyBackRewards(mAssets[i], mAssetToTreasury, mAssetsSellAmount, rewardsAmount);
```

Comment:
mAssets[i] is called more than once, is it worthy to save it as a variable. ? 
mAsset = mAssets[i]

Reply:
mAssets is calldata so it's not an expensive storage variable. Having said that, saving gas is always good so I tried adding `mAsset = mAssets[i];` to see what the difference was. Unfortunately, this resulted in the stack too deep error.

```
CompilerError: Stack too deep, try removing local variables.
   --> contracts/buy-and-make/RevenueSplitBuyBack.sol:243:21:
    |
243 |                     minBassetsAmounts[i],
    |                     ^^^^^^^^^^^^^^^^^
```

I could move part of the logic in the loop out into an internal function to avoid this error but I think we'll loose any gas savings if we do this.

`buyBackRewards` is only called once a week so probably not spending hours saving a small amount of gas.